fx quant: move output obs logic to QuantizeHandler

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #57402 fx quant: remove `find_quants` from convert
* #57399 fx quant: remove unnecessary quants arguments
* #57393 fx quant: remove FixedQParamsOpQuantizeHandler from quantize.py
* #57388 fx quant: move input_output_observed to qhandler
* **#57377 fx quant: move output obs logic to QuantizeHandler**
* #57368 fx quant: readability improvements on observer functions
* #57367 fx quant: clean up nit in insert_observer

Summary:

Moves the logic which determines
1. whether a pattern instance's output should be observed
2. whether a pattern instance's output should be marked as observed based on its inputs
3. whether to ovverride the activation specified in the qconfig

from `quantize.py` to `quantization_patterns.py`.  This makes
the code easier to read and reduces the coupling between `Quantizer`
and `QuantizeHandler` instances.

Note: there are some further cleanups which would be good after this one
- leaving those for future PRs.

Test Plan:

```
python test/test_quantization.py TestQuantizeFxOps
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28126896](https://our.internmc.facebook.com/intern/diff/D28126896)

vkuzo committed:
fx quant: move output obs logic to QuantizeHandler

Summary:

Moves the logic which determines
1. whether a pattern instance's output should be observed
2. whether a pattern instance's output should be marked as observed based on its inputs
3. whether to ovverride the activation specified in the qconfig

from `quantize.py` to `quantization_patterns.py`.  This makes
the code easier to read and reduces the coupling between `Quantizer`
and `QuantizeHandler` instances.

Note: there are some further cleanups which would be good after this one
- leaving those for future PRs.

Test Plan:

```
python test/test_quantization.py TestQuantizeFxOps
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 reviewed (approved):
this is great, thanks!

vkuzo committed:
Update on "fx quant: move output obs logic to QuantizeHandler"


Summary:

Moves the logic which determines
1. whether a pattern instance's output should be observed
2. whether a pattern instance's output should be marked as observed based on its inputs
3. whether to ovverride the activation specified in the qconfig

from `quantize.py` to `quantization_patterns.py`.  This makes
the code easier to read and reduces the coupling between `Quantizer`
and `QuantizeHandler` instances.

Note: there are some further cleanups which would be good after this one
- leaving those for future PRs.

Test Plan:

```
python test/test_quantization.py TestQuantizeFxOps
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28126896](https://our.internmc.facebook.com/intern/diff/D28126896)

[ghstack-poisoned]
